Add free subway transfer logic to base fare estimates

thecristen opened the pull request:
#### Summary of changes
**Asana Ticket:** [Base Fares | Add transfers logic (subway)](https://app.asana.com/0/555089885850811/1169236529454063)

Slightly adjusts the implementation of adding the transfers note, and also introduces logic for identifying subway-subway transfers and accounting for this free transfer when computing the highest one way fare.

Some caveats: there doesn't seem to be an iron-clad way to figure out whether the OTP-generated itinerary contains a real _behind the fare gate_ free subway transfer.


---

Before getting review, please check the following:

* [x] Does frontend functionality render and work correctly in IE?
* [x] Have we load-tested any new pages or internal API endpoints that will receive significant traffic?
* [x] Are interactive elements accessible to screen readers?
* [x] Have you checked for tech debt you can address in the area you're working in?
* [x] If this change involves routes, does it work correctly with pertinent "unusual" routes such as the combined Green Line, Silver Line, Foxboro commuter rail, and single-direction bus routes like the 170?
* [x] Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?

arkadyan commented on apps/fares/lib/transfer.ex:
```diff
@@ -0,0 +1,120 @@
+  Takes a pair of routes and returns true if there might be a transfer between
+  the two, based on the list in @single_ride_valid_transfers
+  """
+  @spec is_maybe_transfer([Route.id_t()]) :: boolean | nil
```
Nit: the [convention](https://hexdocs.pm/elixir/master/naming-conventions.html#trailing-question-mark-foo) in Elixir is to name functions that return booleans with a trailing question mark. (Same thing for `is_free_transfer`.)

thecristen commented on apps/fares/lib/transfer.ex:
```diff
@@ -0,0 +1,120 @@
+  end
+
+  @spec to_fare_atom(fare_atom | Route.id_t() | Route.t()) :: fare_atom
+  defp to_fare_atom(route_or_atom) do
```
Ha, just tried this refactor and it went slightly awry! But I will change the name of the ScheduleView function to better indicate the difference.

ingridpierre commented:
Planning a trip from South Station to Wonderland (for example),  subway transfer logic isn't coming through. Prices appear to still be added 3x.
![image](https://user-images.githubusercontent.com/5806643/83817340-7d8c3c80-a692-11ea-93e7-dee0482f1f28.png)

thecristen commented:
Thank you!! I'd forgotten about the concept of parent stations and platforms sometimes having their own stop ids. Update forthcoming...

thecristen commented:
Okay I don't know why CI keeps failing me (because it passes on my machine, _classic_), but please go ahead and re-review. 

It should now properly handle transfers at large stations, and should be adding the fare properly but many-route multimodal trips like that pictured above! @arkadyan I think I've addressed all your comments but I've refactored a fair bit so this could use another look.

arkadyan commented on apps/fares/lib/transfer.ex:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  @doc "Searches a list of legs for evidence of an in-station subway transfer."
+  @spec is_subway_transfer?([Leg.id_t()]) :: boolean
```
Here's one thing that dialyzer is complaining about: `Leg` doesn't have an `id_t` type.
